src/libsyntax/ext/expand.rs

```diff
@@ -1103,6 +1156,40 @@ pub fn expand_crate(mut cx: ExtCtxt,
     return (ret, cx.syntax_env.names);
 }
 
+// Expands crate using supplied MacroExpander - allows for
+// non-standard expansion behaviour (eg. stepwise)
+pub fn expand_crate_with_expander(expander: &mut MacroExpander,
+                                  user_exts: Vec<NamedSyntaxExtension>,
+                                  mut c: Crate) -> (Crate, HashSet<Name>) {
+    if std_inject::no_core(&c) {
+        expander.cx.crate_root = None;
+    } else if std_inject::no_std(&c) {
+        expander.cx.crate_root = Some("core");
+    } else {
+        expander.cx.crate_root = Some("std");
+    }
+    let ret = {
+        for (name, extension) in user_exts {
+            expander.cx.syntax_env.insert(name, extension);
+        }
+
+        let items = SmallVector::many(c.module.items);
+        expander.load_macros(&items);
+        c.module.items = items.into();
+
+        let err_count = expander.cx.parse_sess.span_diagnostic.err_count();
+        let mut ret = expander.fold_crate(c);
+        ret.exported_macros = expander.cx.exported_macros.clone();
+
+        if expander.cx.parse_sess.span_diagnostic.err_count() > err_count {
+            expander.cx.parse_sess.span_diagnostic.abort_if_errors();
+        }
+
+        ret
+    };
```

Comment:
what is the block for? I don't see an RAII use. Could you comment in the code please?